crates/kitsune_p2p/types/src/transport_pool.rs

```diff
@@ -0,0 +1,257 @@
+    fn handle_debug(&mut self) -> TransportListenerHandlerResult<serde_json::Value> {
+        let out = self
+            .sub_listeners
+            .iter()
+            .map(|(k, v)| {
+                let k = k.to_string();
+                let v = v.debug();
+                async move { TransportResult::Ok((k, v.await?)) }
+            })
+            .collect::<Vec<_>>();
+        Ok(async move {
+            let v = futures::future::try_join_all(out).await?;
+            let m = v
+                .into_iter()
+                .collect::<serde_json::map::Map<String, serde_json::Value>>();
+            Ok(m.into())
+        }
+        .boxed()
+        .into())
+    }
+
+    fn handle_bound_url(&mut self) -> TransportListenerHandlerResult<url2::Url2> {
+        let urls = self
+            .sub_listeners
+            .iter()
+            .map(|(k, v)| {
+                let k = k.to_string();
+                let v = v.bound_url();
+                async move { TransportResult::Ok((k, v.await?)) }
+            })
+            .collect::<Vec<_>>();
+        Ok(async move {
+            let urls = futures::future::try_join_all(urls).await?;
+            let mut out = url2::url2!("kitsune-pool:pool");
+            {
+                let mut query = out.query_pairs_mut();
+                for (k, v) in urls {
+                    query.append_pair(&k, v.as_str());
+                }
+            }
+            Ok(out)
+        }
+        .boxed()
+        .into())
+    }
+
+    fn handle_create_channel(
+        &mut self,
+        url: url2::Url2,
+    ) -> TransportListenerHandlerResult<(url2::Url2, TransportChannelWrite, TransportChannelRead)>
+    {
+        // TODO - right now requiring sub transport scheme to create channel
+        //        would be nice to also accept a pool url && prioritize the
+        //        sub-scheme.
+        let scheme = url.scheme().to_string();
+        match self.sub_listeners.get(&scheme) {
+            None => Err(format!("no sub-transport matching scheme '{}' in pool", scheme).into()),
+            Some(s) => Ok(s.create_channel(url)),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::transport_mem::*;
+    use futures::stream::StreamExt;
+
+    fn test_receiver(mut recv: TransportEventReceiver) {
+        tokio::task::spawn(async move {
+            while let Some(evt) = recv.next().await {
+                match evt {
+                    TransportEvent::IncomingChannel(url, mut write, read) => {
+                        let data = read.read_to_end().await;
+                        let data = format!("echo({}): {}", url, String::from_utf8_lossy(&data),);
+                        write.write_and_close(data.into_bytes()).await?;
+                    }
+                }
+            }
+            TransportResult::Ok(())
+        });
+    }
+
+    #[tokio::test(threaded_scheduler)]
+    async fn it_can_pool_transport() -> TransportResult<()> {
```

Comment:
I can't understand what this test is proving. It looks like you can create two separate transport pools and send a requests to each and get back a response?
But I don't see why there are two pools in this test? It seems like they don't interact.
Maybe you could add a comment explaining the intent of this test?

Reply:
The base mem-transports can talk to each other... if you notice, I'm sending the address for 2 in the request to 1, and visa-versa.

So the two pool transports are talking to each other through their sub mem-transports. - I can add some comments to the test.